### PR TITLE
feat: Add support for mixed routes in the interface

### PR DESCRIPTION
**Pull request by zhongeric**
The new versions of the smart order router (SOR) and routing-api introduce support for a new kind of route called `MixedRoutes`. MixedRoutes are routes which have both V2 and V3 pools in them, ex:

`USDC -[V3]-> ETH -[V2]-> USDT`

We only support returning these routes through the AutoRouter / routingAPI, and so we can pass in a new `protocols` queryParam of `[Protocol.V2, Protocol.V3, Protocol.MIXED]` to direct the auto router to also look for and consider mixedRoutes. Client side, we should still only query for V3 routes using the client side SOR and/or the QuoterV2 fallback.

We also need to edit the `Trade` object we create here to include the new mixed route, in order to generate the correct calldata to submit the swap with.

We also show a new badge label of `V3 + V2` to denote the new mixed routes in the RoutingDiagram

### CODE REVIEW

**Comment by zhongeric** [comment]
TODO: do we want to quote mixedRoutes in `useClientSideTrade` hook?

File: src/components/RoutingDiagram/RoutingDiagram.tsx [hunk 1]
```diff
@@ -73,6 +74,10 @@ const ProtocolBadge = styled(Badge)`
   z-index: ${Z_INDEX.sticky + 1};
 `
 
+const MixedProtocolBadge = styled(ProtocolBadge)`
+  width: 60px;
+`
```
Comment (zhongeric) [inline src/components/RoutingDiagram/RoutingDiagram.tsx:79]: This seems to work to prevent the text from overflowing and line breaking

File: src/state/routing/slice.ts [hunk 2]
```diff
@@ -36,7 +36,7 @@ const DEFAULT_QUERY_PARAMS = {
 export const routingApi = createApi({
   reducerPath: 'routingApi',
   baseQuery: fetchBaseQuery({
-    baseUrl: 'https://api.uniswap.org/v1/',
+    baseUrl: 'https://r6oz6jqnth.execute-api.us-east-2.amazonaws.com/prod',
```
Comment (zhongeric) [inline src/state/routing/slice.ts:39]: ignore this

**Comment by vm** [comment]
@zhongeric can you rebase/merge main?

File: src/hooks/useAutoSlippageTolerance.ts [hunk 3]
```diff
@@ -43,8 +45,17 @@ function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined)
...
+            gas += V3_SWAP_BASE_GAS_ESTIMATE + section.length * V3_SWAP_HOP_GAS_ESTIMATE
+          } else if (section.every((pool) => pool instanceof Pair)) {
+            gas += V2_SWAP_BASE_GAS_ESTIMATE + (section.length - 1) * V2_SWAP_HOP_GAS_ESTIMATE
+          }
+        })
       } else {
-        // TODO: Update with better estimates once we have interleaved routes.
+        // fallback general gas estimation
         gas += V3_SWAP_BASE_GAS_ESTIMATE + route.pools.length * V3_SWAP_HOP_GAS_ESTIMATE
       }
```
Comment (zhongeric) [inline src/hooks/useAutoSlippageTolerance.ts:60]: Chose to keep an `else` case as a fallback, not really needed tho

**Comment by zhongeric** [comment]
Do we want to query for mixedRoutes in the `getClientSideQuote` function as well? See: https://github.com/Uniswap/interface/pull/4181/files#diff-c9fd69a3658e7e36f7ab6b62fd18589fed617dad15c3184bc7335b9e6ad7ee02R68

File: src/hooks/useAutoSlippageTolerance.ts [hunk 4]
```diff
@@ -43,8 +45,19 @@ function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined)
...
+        const sections = partitionMixedRouteByProtocol(route as MixedRoute<Currency, Currency>)
+        let acc = 0
+        sections.forEach((section) => {
+          if (section.every((pool) => pool instanceof Pool)) {
+            acc += V3_SWAP_BASE_GAS_ESTIMATE + section.length * V3_SWAP_HOP_GAS_ESTIMATE
+          } else if (section.every((pool) => pool instanceof Pair)) {
+            acc += V2_SWAP_BASE_GAS_ESTIMATE + (section.length - 1) * V2_SWAP_HOP_GAS_ESTIMATE
+          }
+        })
+        gas += acc
```
Comment (zhongeric) [inline src/hooks/useAutoSlippageTolerance.ts:58]: Had to use another acc var here because of esLint no-func-loop

File: src/state/routing/slice.ts [hunk 5]
```diff
@@ -36,7 +37,7 @@ const DEFAULT_QUERY_PARAMS = {
 export const routingApi = createApi({
   reducerPath: 'routingApi',
   baseQuery: fetchBaseQuery({
-    baseUrl: 'https://api.uniswap.org/v1/',
+    baseUrl: 'https://beta.api.uniswap.org/v1/',
```
Comment (snreynolds) [inline src/state/routing/slice.ts:40]: @zhongeric this should still point to prod. we will test on beta this morning and deploy to prod before releasing interface

Comment (zhongeric) [reply src/state/routing/slice.ts]: ok sg

File: src/state/routing/slice.ts [hunk 6]
```diff
@@ -28,7 +28,6 @@ const protocols: Protocol[] = [Protocol.V2, Protocol.V3, Protocol.MIXED]
 
 const DEFAULT_QUERY_PARAMS = {
   protocols: protocols.map((p) => p.toLowerCase()).join(','),
-  forceMixedRoutes: true,
```
Comment (vm) [inline src/state/routing/slice.ts]: this part should be done in a separate PR

Comment (vm) [reply src/state/routing/slice.ts]: that way we know what we can minimally revert if issues arise

Comment (zhongeric) [reply src/state/routing/slice.ts]: makes sense, will cut

Comment (zhongeric) [reply src/state/routing/slice.ts]: #4326

Comment (vm) [reply src/state/routing/slice.ts]: I think all that PR needs to be is `forceMixedRoutes: true`?

File: src/hooks/useAutoSlippageTolerance.ts [hunk 7]
```diff
@@ -43,8 +45,19 @@ function guesstimateGas(trade: Trade<Currency, Currency, TradeType> | undefined)
         // V3 gas costs scale on initialized ticks being crossed, but we don't have that data here.
         // We bake in some tick crossings into the base 100k cost.
         gas += V3_SWAP_BASE_GAS_ESTIMATE + route.pools.length * V3_SWAP_HOP_GAS_ESTIMATE
+      } else if (route.protocol === Protocol.MIXED) {
+        const sections = partitionMixedRouteByProtocol(route as MixedRoute<Currency, Currency>)
+        let acc = 0
```
Comment (zzmp) [inline src/hooks/useAutoSlippageTolerance.ts:50]: `acc` is unnecssary - you should instead just have L53 and L55 use `gas +=` to be more explicit about what you're adding up.

You should also use `reduce` to make it more clear what is happening in the code:

    gas += sections.reduce((gas, section) => {
      if (section.every((pool) => pool instanceof Pool)) {
        return gas + V3_SWAP_BASE_GAS_ESTIMATE + section.length * V3_SWAP_HOP_GAS_ESTIMATE
      } else if (section.every((pool) => pool instanceof Pair)) {
        return gas + V2_SWAP_BASE_GAS_ESTIMATE + (section.length - 1) * V2_SWAP_HOP_GAS_ESTIMATE
      } else {
        console.warn('Invalid section')
        return gas
      }
    })

File: src/state/routing/utils.ts [hunk 8]
```diff
@@ -97,6 +112,19 @@ const parsePair = ({ reserve0, reserve1 }: V2PoolInRoute): Pair =>
...
+function isV3Route(route: (V3PoolInRoute | V2PoolInRoute)[]): route is V3PoolInRoute[] {
+  return route.every((pool) => pool.type === 'v3-pool')
+}
+
+function isV2Route(route: (V3PoolInRoute | V2PoolInRoute)[]): route is V2PoolInRoute[] {
+  return route.every((pool) => pool.type === 'v2-pool')
+}
+
+function isMixedRoute(route: (V3PoolInRoute | V2PoolInRoute)[]): route is (V3PoolInRoute | V2PoolInRoute)[] {
+  /// Must have at least one V3 pool and one V2 pool
```
Comment (zzmp) [inline src/state/routing/utils.ts:128]: This isn't comprehensive. Should you also check that every route is either V3 or V2?

File: src/state/routing/utils.ts [hunk 9]
```diff
@@ -71,6 +79,13 @@ export function transformRoutesToTrade<TTradeType extends TradeType>(
       route
         ?.filter((r): r is typeof route[0] & { routev3: NonNullable<typeof route[0]['routev3']> } => r.routev3 !== null)
         .map(({ routev3, inputAmount, outputAmount }) => ({ routev3, inputAmount, outputAmount })) ?? [],
+    mixedRoutes:
```
Comment (zzmp) [inline src/state/routing/utils.ts:82]: Where is this used?

**Review by zzmp — CHANGES_REQUESTED** [review]

File: src/hooks/useAutoSlippageTolerance.ts (hunk 7 above)
Comment (zhongeric) [reply src/hooks/useAutoSlippageTolerance.ts:50]: I like the .reduce

File: src/state/routing/utils.ts [hunk 10]
```diff
@@ -41,8 +42,15 @@ export function computeRoutes(
       }
 
       return {
```
Comment (zzmp) [inline src/state/routing/utils.ts:44]: This requires a lot of looping that isn't necessary. Could you improve it by doing something like this?

    getRouteProtocol(route: (V3PoolInRoute | V2PoolInRoute)[]): Protocol {
      if (route.every((pool) => pool.type === 'v2-pool')) return Protocol.V2
      if (route.every((pool) => pool.type === 'v3-pool')) return Protocol.V3
      return Protocol.MIXED
    }

You could optimize it more but this seems sufficient. The old `isV2Route`-style functions were type guards, but the `genericPoolPairParser` makes it so that you no longer need to cast to a type before parsing the route.

Comment (zhongeric) [reply src/state/routing/utils.ts:44]: Yeah this is a lot more elegant. This also solves the comment above about the mixed routes not exactly being comprehensive

File: src/state/routing/utils.ts (hunk 8 above)
Comment (zhongeric) [reply src/state/routing/utils.ts:128]: Should be resolved by new logic in `getRouteProtocol`

File: src/state/routing/utils.ts (hunk 9 above)
Comment (zhongeric) [reply src/state/routing/utils.ts:87]: used to build the `InterfaceTrade` which is returned by `useRoutingAPITrade()` and then in `useBestTrade()`

File: src/state/routing/utils.ts [hunk 11]
```diff
@@ -41,16 +41,21 @@ export function computeRoutes(
         throw new Error('Expected both amountIn and amountOut to be present')
       }
 
+      const routeProtocol: Protocol = getRouteProtocol(route)
```
Comment (zzmp) [inline src/state/routing/utils.ts:44]: rm `: Protocol`, the typing is unneeded here. It can be inferred from the return value of `getRouteProtocol`.

**Review by zzmp — APPROVED** [review]
